fix(mouse): check that mouse support is enabled for visual mode

Exagone313 opened the pull request:
Fixes #15007

---

If mouse support is not enabled for visual mode, dragging the mouse in another mode should not begin a selection.

This is a regression introduced in 6db86cb2d3d4ca152f156dc07362f8796150fae0.

---

For the record, the check is still in place in latest vim:

https://github.com/vim/vim/blob/056678184f679c2989b73bd48eda112f3c79a62f/src/mouse.c#L638-L639

https://github.com/vim/vim/blob/056678184f679c2989b73bd48eda112f3c79a62f/src/mouse.c#L659-L660

`mouse_has` definition:

https://github.com/vim/vim/blob/056678184f679c2989b73bd48eda112f3c79a62f/src/mouse.c#L1432-L1456

https://github.com/neovim/neovim/blob/d0668b36a3e2d0683059baead45bea27e2358e9c/src/nvim/mouse.c#L555-L575

Exagone313 commented:
Failed tests were added in 810da1a7023ec87e3fa644499a655740c01baaf7 and 28ac6c00e6f61d40f082af3ae3db380d7951acee, both commits were made after the regression was introduced.

bfredl commented:
Ugh. so `'mouse'` in vim does not only control _if_ the mouse is captured, but also what it end up _doing_. The vim documentation even only mentions the former.. I am tempted to just ignore this "feature" of vim and call the current behavior the correct one for nvim, so the advice at mouse_on/mouse_off still applies.

Exagone313 commented:
@bfredl, I rely on this weird behavior, an alternative would be to make it optional. Though `help dev` says `Backwards compatibility is a feature.`.

zeertzjq commented on src/nvim/normal.c:
```diff
@@ -1468,6 +1468,22 @@ static void move_tab_to_mouse(void)
   }
 }
 
+// Check that mouse support is enabled for visual mode
+static int mouse_supports_visual(void)
```
```suggestion
/// Check that mouse support is enabled for visual mode
static bool mouse_supports_visual(void)
```

zeertzjq commented on src/nvim/normal.c:
```diff
@@ -1468,6 +1468,22 @@ static void move_tab_to_mouse(void)
+// Check that mouse support is enabled for visual mode
+static int mouse_supports_visual(void)
+{
+  for (char_u *p = p_mouse; *p; ++p)
+    switch (*p) {
+    case 'a':
+        return true;
+      break;
+    case MOUSE_HELP: if (curbuf->b_help)
+        return true;
+      break;
+    default: if (MOUSE_VISUAL == *p) return true; break;
+    }
```
```suggestion
  for (char_u *p = p_mouse; *p; p++) {
    switch (*p) {
    case 'a':
        return true;
      break;
    case MOUSE_HELP: if (curbuf->b_help)
        return true;
      break;
    default: if (MOUSE_VISUAL == *p) return true; break;
    }
  }
```

bfredl commented:
@Exagone313 Forward compatibility is also a goal. neovim wants to support full mouse input in GUI even if the 'mouse' option is set to prefer terminal emulators own mouse over nvim's. 

But I think there is a reasonable way to opt-in: enable this behavior _only_ if 'mouse' contains `n` but not `v`. Then the default of `mouse=` is not affected. What do you think?

Exagone313 commented:
@bfredl, I have edited the function to return true when mouse is empty, is it ok or should I explicitly check the case where it contains `n` but not `v`?

bfredl commented:
Let's try that for now, it is probably enough.

Exagone313 commented:
If going into that other direction, I think the function would be:
```c
/// Check that mouse support is enabled for visual mode
static bool mouse_supports_visual(void)
{
  return p_mouse[0] == '\0' || vim_strchr(p_mouse, 'a') != NULL
    || vim_strchr(p_mouse, 'v') != NULL;
}
```

Exagone313 commented:
The test suite failing on FreeBSD works locally on Linux and seems to work on OpenBSD too.

Exagone313 commented on src/nvim/normal.c:
```diff
@@ -1468,6 +1468,32 @@ static void move_tab_to_mouse(void)
+/// Check that mouse support is enabled for visual mode
+static bool mouse_supports_visual(void)
+{
+  if (!*p_mouse) {
```
If wanted, I can do this:
```diff
-  if (!*p_mouse) {
+  if (p_mouse[0] == '\0') {
```
I've seen code doing that in neovim.

bfredl replied:
either one is fine.

Exagone313 commented:
I noticed that this patch doesn't fix the whole behavior. Double click is also to be considered.

justinmk commented on src/nvim/normal.c:
```diff
@@ -1468,6 +1468,32 @@ static void move_tab_to_mouse(void)
 }
 
+/// Check that mouse support is enabled for visual mode
+static bool mouse_supports_visual(void)
```
this is rather specific and looks mostly copy-pasted from `ui_check_mouse`. Any way to extract the logic from `ui_check_mouse` into something like `mouse_is_mode_enabled` ?

Exagone313 replied:
Yes, and previous code was seperating this logic in a separated function, but the refactoring that happened in 6db86cb2d3d4ca152f156dc07362f8796150fae0 moved it (since only one call was needed).

justinmk replied:
https://github.com/neovim/neovim/commit/6db86cb2d3d4ca152f156dc07362f8796150fae0 inlined `mouse_has()` because there didn't seem to be a need for it. If there is now a need for it (or something like it), then it can be restored... that seems preferable to copy-pasting logic.

justinmk commented:
Thanks for updating the PR, looks like my suggestion is still pending. 

Also do you need help writing a test? [test/functional/example_spec.lua](https://github.com/neovim/neovim/blob/master/test/functional/example_spec.lua) for example. You can [run the test](https://github.com/neovim/neovim/blob/master/test/README.md#filter-by-file) like this:

    TEST_FILE=test/functional/example_spec.lua make functionaltest

Exagone313 commented:
This PR is still on my mind, as I self build Neovim. But there are other fixes needed, for example for double-click.

I just need to find some time to continue working on this.

Exagone313 commented:
I fixed the issue with double click.

I think I need to add relevant tests in `test/functional/ui/mouse_spec.lua` as it contains related test cases already.

I have read the `ui_check_mouse` function again and I'm not sure how to extract and re-use the logic.

Exagone313 committed:
fix(mouse): check that mouse support is enabled for visual mode

If mouse support is not enabled for visual mode, dragging the mouse or
double clicking in another mode should not begin a selection.

This is a regression introduced in
6db86cb2d3d4ca152f156dc07362f8796150fae0.

Exagone313 commented:
@justinmk I need some help on writing functional tests. 

I have trouble understanding how `screen:expect` works.  I don't understand the test `single left click moves cursor` in `ui/mouse_spec.lua`: shouldn't it enter visual mode when moving the mouse (there are no `LeftRelease`)? It sounds like this test should have caught the current bug I'm dealing with. The behavior seems different in the test `double left click enters visual mode` (where `LeftRelease` is used).

Thanks.

zeertzjq commented:
You need to use `<LeftDrag>` to move the mouse.

Exagone313 commented:
I'm no longer interested in working on this. I have updated my configuration to use `mouse=a`.
